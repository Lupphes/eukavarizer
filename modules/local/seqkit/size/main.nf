process SEQKIT_SIZE {
    tag "$meta.id"
    label 'process_low'

    input:
    tuple val(meta), path(reads)

    output:
    tuple val(meta), path(reads), path("*.txt"), emit: median_bp
    path "versions.yml"                        , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    # Compute quick average read length from first 1000 reads (4000 lines)
    ( zcat ${reads.collect { "\"${it}\"" }.join(' ')} || true ) | \\
        head -n 4000 | \\
        awk 'NR % 4 == 2 { total += length(\$0); count++ } END { if (count > 0) printf "%d\\n", total / count; else print 0 }' > ${prefix}_median_size.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        awk: \$(awk --version 2>&1 | head -n1 | sed 's/.*Awk //; s/,.*//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo "0" > ${prefix}_median_size.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        awk: \$(awk --version 2>&1 | head -n1 | sed 's/.*Awk //; s/,.*//')
    END_VERSIONS
    """
}
