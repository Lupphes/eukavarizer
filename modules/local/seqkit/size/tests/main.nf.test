nextflow_process {

    name "Test Process SEQKIT_SIZE"
    script "../main.nf"
    process "SEQKIT_SIZE"
    tag "seqkit"
    tag "seqkit/size"
    tag "modules"
    tag "modules_local"

    test("Single-end FASTQ - success") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id: 'test', single_end: true ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz')
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.median_bp.size() == 1 },
                { assert process.out.versions }
            )
        }
    }

    test("Paired-end FASTQ - success") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id: 'test', single_end: false ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz'),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz')
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.median_bp.size() == 1 },
                { assert process.out.versions }
            )
        }
    }

    test("Empty FASTQ - should handle gracefully") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id: 'empty', single_end: true ],
                    file(params.modules_testdata_base_path + 'generic/txt/hello.txt')
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.versions }
            )
        }
    }
}
