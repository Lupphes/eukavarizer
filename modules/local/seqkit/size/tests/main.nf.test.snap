{
    "Single-end FASTQ - success": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test_1.fastq.gz:md5,4161df271f9bfcd25d5845a1e220dbec",
                        "test_median_size.txt:md5,26e41e2047e361fd54e4d19ec39375eb"
                    ]
                ],
                "1": [
                    "versions.yml:md5,43e3eb957ae70f2530578840819a7314"
                ],
                "median_bp": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test_1.fastq.gz:md5,4161df271f9bfcd25d5845a1e220dbec",
                        "test_median_size.txt:md5,26e41e2047e361fd54e4d19ec39375eb"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,43e3eb957ae70f2530578840819a7314"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-20T22:36:24.389301335"
    },
    "Empty FASTQ - should handle gracefully": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "empty",
                            "single_end": true
                        },
                        "hello.txt:md5,e59ff97941044f85df5297e1c302d260",
                        "empty_median_size.txt:md5,897316929176464ebc9ad085f31e7284"
                    ]
                ],
                "1": [
                    "versions.yml:md5,43e3eb957ae70f2530578840819a7314"
                ],
                "median_bp": [
                    [
                        {
                            "id": "empty",
                            "single_end": true
                        },
                        "hello.txt:md5,e59ff97941044f85df5297e1c302d260",
                        "empty_median_size.txt:md5,897316929176464ebc9ad085f31e7284"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,43e3eb957ae70f2530578840819a7314"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-20T22:36:40.872782337"
    },
    "Paired-end FASTQ - success": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test_1.fastq.gz:md5,4161df271f9bfcd25d5845a1e220dbec",
                            "test_2.fastq.gz:md5,2ebae722295ea66d84075a3b042e2b42"
                        ],
                        "test_median_size.txt:md5,26e41e2047e361fd54e4d19ec39375eb"
                    ]
                ],
                "1": [
                    "versions.yml:md5,43e3eb957ae70f2530578840819a7314"
                ],
                "median_bp": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test_1.fastq.gz:md5,4161df271f9bfcd25d5845a1e220dbec",
                            "test_2.fastq.gz:md5,2ebae722295ea66d84075a3b042e2b42"
                        ],
                        "test_median_size.txt:md5,26e41e2047e361fd54e4d19ec39375eb"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,43e3eb957ae70f2530578840819a7314"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-20T22:36:32.530027386"
    }
}